File: sdks/python/croupier/trace.py
"""OTel trace 传播（一期·无依赖版）。

平台在 invoke 的 handler context JSON（即 metadata 序列化）里携带 W3C
``traceparent`` 与冗余明文 ``trace_id``（见
docs/architecture/sdk-otel-propagation.md）。本模块提供读取辅助——游戏方
据此做日志关联，或接入自己的 OTel 体系（SDK 不内置 exporter/自动埋点）。
"""

from __future__ import annotations

import json
from typing import Any, Dict

METADATA_TRACEPARENT = "traceparent"
METADATA_TRACE_ID = "trace_id"


def trace_parent_from_context(context_json: str) -> str:
    """从 handler context JSON 提取 W3C traceparent（无则空串）。"""
    return _field(context_json, METADATA_TRACEPARENT)


def trace_id_from_context(context_json: str) -> str:
    """从 handler context JSON 提取 trace_id 明文（无则空串）。

    可直接拼进游戏方日志/错误信息，去 Jaeger/Grafana 查询整条链路。
    """
    return _field(context_json, METADATA_TRACE_ID)


def _field(context_json: str, key: str) -> str:
    if not context_json:
        return ""
    try:
        parsed: Any = json.loads(context_json)
    except (TypeError, ValueError):
        return ""
    if not isinstance(parsed, dict):
        return ""
    value = parsed.get(key)
    if isinstance(value, str):
        return value.strip()
    return ""


__all__ = [
    "METADATA_TRACEPARENT",
    "METADATA_TRACE_ID",
    "trace_parent_from_context",
    "trace_id_from_context",
]

File: sdks/go/pkg/croupier/trace.go
package croupier

import (
	"context"
	"strings"
)

// OTel trace 传播（一期·无依赖版）：平台在 invoke metadata 携带 W3C
// traceparent 与冗余明文 trace_id（见 docs/architecture/sdk-otel-propagation.md）。
// SDK 职责仅是把这些值提取进 handler 上下文——游戏方据此做日志关联，
// 或自行接入其 OTel 体系（SDK 不内置 exporter/自动埋点）。

// metadata 约定键（与平台 telemetry.InjectContext 一致）。
const (
	MetadataTraceparent = "traceparent"
	MetadataTraceID     = "trace_id"
)

type traceContextKey struct{}

type traceContext struct {
	traceparent string
	traceID     string
}

// WithTraceMetadata 把 invoke metadata 中的 trace 字段写入 ctx（无则原样返回）。
func WithTraceMetadata(ctx context.Context, meta map[string]string) context.Context {
	if ctx == nil {
		ctx = context.Background()
	}
	tp := strings.TrimSpace(meta[MetadataTraceparent])
	tid := strings.TrimSpace(meta[MetadataTraceID])
	if tp == "" && tid == "" {
		return ctx
	}
	return context.WithValue(ctx, traceContextKey{}, traceContext{traceparent: tp, traceID: tid})
}

// TraceParentFromContext 返回当前请求的 W3C traceparent（无则空串）。
func TraceParentFromContext(ctx context.Context) string {
	if tc, ok := ctx.Value(traceContextKey{}).(traceContext); ok {
		return tc.traceparent
	}
	return ""
}

// TraceIDFromContext 返回当前请求的 trace_id 明文（无则空串）。
// 可直接拼进游戏方日志/错误信息，去 Jaeger/Grafana 查询整条链路。
func TraceIDFromContext(ctx context.Context) string {
	if tc, ok := ctx.Value(traceContextKey{}).(traceContext); ok {
		return tc.traceID
	}
	return ""
}
